### PR TITLE
[analyzer] Report violations of the "returns_nonnull" attribute

### PR DESCRIPTION
**Pull request by necto**
Make sure code respects the GNU-extension `__attribute__((returns_nonnull))`.

Extend the NullabilityChecker to check that a function returns_nonnull does not return a nullptr.

This commit also reverts an old hack introduced by 49bd58f1ebe28d97e4949e9c757bc5dfd8b2d72f
because it is no longer needed

CPP-4741

### CODE REVIEW

File: clang/test/Analysis/nullability.c
Comment (NagyDonat) [inline clang/test/Analysis/nullability.c:53]: Consider adding a variant of this where a passthrough function is used in a situation where it's known to return null.
```cpp
__attribute__((returns_nonnull)) int *passthrough2(int *p) {
  return p;
}
void call_with_null(void) {
  passthrough2(NULL);
}
```

File: clang/test/Analysis/nullability.c
Comment (NagyDonat) [inline clang/test/Analysis/nullability.c:43]:
```suggestion
    // expected-warning@-1 {{REACHABLE}}
    // TODO: This warning is a false positive, according to the contract of
    // produce_nonnull_ptr, x cannot be null.
```
(Just bikeshedding.)

By the way, do you have plans to eliminate this false positive (perhaps in a follow-up commit)?

File: clang/test/Analysis/nullability.c
Comment (NagyDonat) [inline clang/test/Analysis/nullability.c:1]: Consider eliminating the use of  `-Wno-deprecated-non-prototype` by specifiying the parameter lists of functions. (Note that this is C, so you need `return_t funcname(void)` to specify a function that takes no arguments.)

File: clang/lib/StaticAnalyzer/Checkers/NullabilityChecker.cpp
Comment (NagyDonat) [inline clang/lib/StaticAnalyzer/Checkers/NullabilityChecker.cpp:695]: Why isn't this logic added as a part of `getNullabilityAnnotation`? (I'm just curious, I can easily imagine that there is a good reason for that.)

**Review by NagyDonat — COMMENTED** [review]
Thanks for this improvement, I'm really happy to see this! (I added some inline comments, but they are all minor and tangential.)   

Right now all the `nullability.*` checkers are marked as (ObjC) in [the documentation](https://clang.llvm.org/docs/analyzer/checkers.html#nullability), because (as far as I know) previously they were only supporting the Objective-C style `_Nullable` and `_Nonnull` annotations.

However, as your change shows, the same "backend" can be used to handle different kinds of nullability annotations, which would make these checkers available for more users.

Please update the documentation of the affected checkers (in `clang/docs/analyzer/checkers.rst`) to mark that they are no longer limited to Objective-C.

File: clang/test/Analysis/nullability.c
Comment (necto) [reply clang/test/Analysis/nullability.c]: We've entered it as a Jira ticket: https://sonarsource.atlassian.net/browse/CPP-5614
But we have no immediate plans to implement it.

File: clang/test/Analysis/nullability.c
Comment (necto) [reply clang/test/Analysis/nullability.c]: Done in a9c489f683fc

File: clang/test/Analysis/nullability.c
Comment (necto) [reply clang/test/Analysis/nullability.c:56]: Added in 3d5e750aec0d 
However, it did not trigger because of an explicit suppression of the reports in inlined functions:
https://github.com/llvm/llvm-project/commit/49bd58f1ebe28d97e4949e9c757bc5dfd8b2d72f

I believe the reasoning for that suppression no longer holds, so I lifted the suppression: 598c574e0703
Do you agree? If so, do you think it should be done in a separate change set?

File: clang/lib/StaticAnalyzer/Checkers/NullabilityChecker.cpp
Comment (necto) [reply clang/lib/StaticAnalyzer/Checkers/NullabilityChecker.cpp:695]: `getNullabilityAnnotation` is concerned only with the type. It is used both for return types and function parameters by multiple checkers.
Handling the function annotation there would mean I have to pass the function declaration (which might not be available in the client of the API), which is only relevant in one case (and maybe another couple in TrustNonnullChecker). That means I would need to complicate the function definition and potentially increase the PR footprint by touching all call sites.

**Comment by necto** [comment]
> Please update the documentation of the affected checkers (in clang/docs/analyzer/checkers.rst) to mark that they are no longer limited to Objective-C.

Updated in 2d2ab3185a00

File: clang/docs/analyzer/checkers.rst
Comment (NagyDonat) [inline clang/docs/analyzer/checkers.rst:592]:
```suggestion
nullability.NullReturnedFromNonnull (C, C++, ObjC)
""""""""""""""""""""""""""""""""""""""""""""""""""
```
This is how other checkers are marked. (Do I understand correctly that `__attribute__((returns_nonnull))` is relevant both under C and C++?)

File: clang/test/Analysis/nullability.c
Comment (NagyDonat) [inline clang/test/Analysis/nullability.c:59]: Is this declaration used somehow?

**Review by NagyDonat — COMMENTED** [review]
Thanks for the updates!

File: clang/docs/analyzer/checkers.rst
Comment (necto) [reply clang/docs/analyzer/checkers.rst]: Yes, it is relevant for both. Thanks for the fix! Applied

File: clang/test/Analysis/nullability.c
Comment (necto) [reply clang/test/Analysis/nullability.c]: It is not used, indeed; thanks for spotting it!
It was relevant in one of the working versions of the test case, but not in the committed one. Removed in 41956123

File: clang/lib/StaticAnalyzer/Checkers/NullabilityChecker.cpp
Comment (NagyDonat) [reply clang/lib/StaticAnalyzer/Checkers/NullabilityChecker.cpp:695]: I see, thanks for the clarification.

File: clang/test/Analysis/nullability.c
Comment (NagyDonat) [reply clang/test/Analysis/nullability.c:56]: If the tests within `nullability.mm` still pass (expect for the over-suppression one), then there is no need to preserve that suppression.

I'd say that it's OK to lift the suppression within this commit, but add a remark like "This commit also reverts an old hack introduced by 49bd58f1ebe28d97e4949e9c757bc5dfd8b2d72f because it is no longer needed." to the commit message. (Perhaps putting it into a separate commit would be _slightly_ more elegant, but I don't think that it would produce real practical benefits.)

**Review by NagyDonat — APPROVED** [review]
Looks good to me now, thanks :smile:. Feel free to merge this / I can help with merging if you don't have commit access yet.

**Pull request merged**
